File: app/src/main/java/com/chooloo/www/callmanager/util/ContactUtils.java
package com.chooloo.www.callmanager.util;

import android.app.Activity;
import android.content.ContentUris;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.provider.ContactsContract;
import android.telephony.PhoneNumberUtils;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.chooloo.www.callmanager.cursorloader.ContactLookupCursorLoader;
import com.chooloo.www.callmanager.entity.Contact;

import static android.Manifest.permission.READ_CONTACTS;
import static com.chooloo.www.callmanager.util.PermissionUtils.checkPermission;

public class ContactUtils {

    public static Contact lookupContact(@NonNull Context context, @NonNull String phoneNumber) {
        if (!checkPermission(context, READ_CONTACTS, true)) {
            return Contact.UNKNOWN;
        }
        return ContactLookupCursorLoader.lookupContact(context, phoneNumber);
    }

    public static void openContact(@NonNull Activity activity, @NonNull Contact contact) {
        try {
            Intent intent = new Intent(Intent.ACTION_VIEW);
            intent.setData(Uri.withAppendedPath(ContactsContract.Contacts.CONTENT_URI, String.valueOf(contact.getContactId())));
            activity.startActivity(intent);
        } catch (Exception e) {
            Toast.makeText(activity, "Oops there was a problem trying to open the contact :(", Toast.LENGTH_SHORT).show();
            e.printStackTrace();
        }
    }

    public static void addContact(@NonNull Activity activity, @NonNull Contact contact) {
        try {
            Intent intent = new Intent(Intent.ACTION_INSERT);
            intent.setType(ContactsContract.Contacts.CONTENT_TYPE);
            intent.putExtra(ContactsContract.Intents.Insert.PHONE, contact.getNumber());
            activity.startActivity(intent);
        } catch (Exception e) {
            Toast.makeText(activity, "Couldn't open add contact dialog", Toast.LENGTH_LONG).show();
            e.printStackTrace();
        }
    }

    public static void editContact(@NonNull Activity activity, @NonNull Contact contact) {
        try {
            Intent intent = new Intent(Intent.ACTION_EDIT, ContactsContract.Contacts.CONTENT_URI);
            intent.setData(ContentUris.withAppendedId(ContactsContract.Contacts.CONTENT_URI, contact.getContactId()));
            activity.startActivity(intent);
        } catch (Exception e) {
            Toast.makeText(activity, "Oops there was a problem trying to open the contact :(", Toast.LENGTH_SHORT).show();
            e.printStackTrace();
        }
    }

    public static void deleteContact(@NonNull Activity activity, @NonNull Contact contact) {
        try {
            Uri uri = Uri.withAppendedPath(ContactsContract.Contacts.CONTENT_URI, Long.toString(contact.getContactId()));
            activity.getContentResolver().delete(uri, null, null);
            Toast.makeText(activity, "Contact Deleted", Toast.LENGTH_LONG).show();
        } catch (Exception e) {
            Toast.makeText(activity, "Contact couldn't be deleted", Toast.LENGTH_LONG).show();
            e.printStackTrace();
        }
    }

    public static void smsContact(@NonNull Activity activity, @NonNull Contact contact) {
        Uri uri = Uri.parse(String.format("smsto:%s", PhoneNumberUtils.normalizeNumber(contact.getNumber())));
        Intent smsIntent = new Intent(Intent.ACTION_SENDTO, uri);
        activity.startActivity(smsIntent);
    }
}

File: app/src/main/java/com/chooloo/www/callmanager/util/ContactUtils.kt
package com.chooloo.www.callmanager.util

import android.Manifest.permission
import android.app.Activity
import android.content.ContentUris
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.provider.ContactsContract
import android.telephony.PhoneNumberUtils
import com.chooloo.www.callmanager.cursorloader.ContactLookupCursorLoader
import com.chooloo.www.callmanager.entity.Contact
import com.chooloo.www.callmanager.ui.base.BaseActivity

object ContactUtils {
    const val PERMISSION_RC_WRITE_CONTACTS = 1

    @JvmStatic
    fun lookupContact(context: Context, phoneNumber: String): Contact {
        return if (PermissionUtils.checkPermission(context, permission.READ_CONTACTS, true)) {
            ContactLookupCursorLoader.lookupContact(context, phoneNumber)
        } else {
            Contact.UNKNOWN
        }
    }

    @JvmStatic
    fun openContact(activity: BaseActivity, contact: Contact) {
        activity.startActivity(Intent(Intent.ACTION_VIEW).apply {
            data = Uri.withAppendedPath(ContactsContract.Contacts.CONTENT_URI, contact.contactId.toString())
        })
    }

    @JvmStatic
    fun addContact(activity: BaseActivity, contact: Contact) {
        activity.startActivity(Intent(Intent.ACTION_INSERT).apply {
            type = ContactsContract.Contacts.CONTENT_TYPE
            putExtra(ContactsContract.Intents.Insert.PHONE, contact.number)
        })
    }

    @JvmStatic
    fun editContact(activity: BaseActivity, contact: Contact) {
        activity.startActivity(Intent(Intent.ACTION_EDIT, ContactsContract.Contacts.CONTENT_URI).apply {
            data = ContentUris.withAppendedId(ContactsContract.Contacts.CONTENT_URI, contact.contactId!!)
        })
    }

    @JvmStatic
    fun deleteContact(activity: BaseActivity, contact: Contact) {
        if (activity.hasPermission(permission.WRITE_CONTACTS)) {
            val uri = Uri.withAppendedPath(ContactsContract.Contacts.CONTENT_URI, java.lang.Long.toString(contact.contactId!!))
            activity.contentResolver.delete(uri, null, null)
            activity.showMessage("Contact Deleted")
        } else {
            activity.showError("Give me a permission and try again")
            activity.askForPermission(permission.WRITE_CONTACTS, PERMISSION_RC_WRITE_CONTACTS)
        }
    }

    @JvmStatic
    fun smsContact(activity: Activity, contact: Contact) {
        val uri = Uri.parse(String.format("smsto:%s", PhoneNumberUtils.normalizeNumber(contact.number)))
        activity.startActivity(Intent(Intent.ACTION_SENDTO, uri))
    }
}